manifests/virtualenv.pp

```diff
@@ -134,12 +137,12 @@
     $pip_flags = "${pypi_index} ${proxy_flag} ${pip_args}"
 
     exec { "python_virtualenv_${venv_dir}":
-      command     => "true ${proxy_command} && ${virtualenv_cmd} ${system_pkgs_flag} -p ${python} ${venv_dir} && ${pip_cmd} --log ${venv_dir}/pip.log install ${pip_flags} --upgrade pip && ${pip_cmd} install ${pip_flags} --upgrade ${distribute_pkg}",
+      command     => "${virtualenv_cmd} ${system_pkgs_flag} -p ${python} ${venv_dir} && ${pip_cmd} --log ${venv_dir}/pip.log install ${pip_flags} --upgrade pip && ${pip_cmd} install ${pip_flags} --upgrade ${distribute_pkg}",
       user        => $owner,
       creates     => "${venv_dir}/bin/activate",
       path        => $_path,
       cwd         => '/tmp',
-      environment => $environment,
+      environment => $environment.union($proxy_environ),
```

Comment:
What would happen if `$proxy_environ` and `$environment` contained different values for `HTTP_PROXY`/`HTTPS_PROXY`?  How would we write a test to find out?

Reply:
Writing the tests shouldn't be too complicated.  You can set any combination of `environment` and `proxy` parameters and then assert the `environment` of the `exec`.  eg. (completely untested, incomplete and probably contains typos etc!)

```ruby
context 'when `proxy` set' do
  let(:title) { '/opt/env1' }
  let(:params) { { proxy: 'http://example.com:3128' } }
  it { is_expected.to contain_exec('python_virtualenv_/opt/env1').with_environment(['HTTP_PROXY=http://example.com:3128','HTTPS_PROXY=http://example.com:3128']) }
end

context 'when `proxy` and `environment` have conflicting parameters' do
  let(:title) { '/opt/env1' }
  let(:params) do
    {
      proxy: 'http://example.com:3128',
      environment: [ 'HTTP_PROXY=http://example.com:8080' ]
    }
  end
  it { is_expected.to contain_exec('python_virtualenv_/opt/env1').with_environment(['HTTP_PROXY=http://example.com:3128','HTTPS_PROXY=http://example.com:3128']) }
end
# etc.
```

I think that second example will fail though.  `union` will not deduplicate the conflicting `HTTP_PROXY=` strings.  If you can add tests like this, I can help with a fix for replacing `union` (assuming the test *does* fail).  Hope this helps.

Reply:
Well, `union` on't deduplicate environment strings, but presumably the shell will.

Reply:
Test added; let's see if it passes.